passes/cmds/wrapcell.cc

```diff
@@ -0,0 +1,170 @@
+		log("exists, it is reused.\n");
+		log("\n");
+		log("    -setattr <attribute-name>\n");
+		log("        set the given boolean attribute on each created wrapper module\n");
+		log("\n");
+		log("    -formatattr <attribute-name> <format>\n");
+		log("        set a string attribute on the created wrapper module by substituting\n");
+		log("        parameter values into <format>\n");
+		log("\n");
+		log("Currently this command only supports wrapping internal cell types.\n");
+		log("\n");
+	}
+
+	void execute(std::vector<std::string> args, Design *d) override
+	{
+		log_header(d, "Executing WRAPCELL pass. (wrap selected cells)\n");
+
+		struct AttrRule {
+			IdString name;
+			std::string value_fmt;
+
+			AttrRule(IdString name, std::string value_fmt)
+				: name(name), value_fmt(value_fmt) {}
+		};
+		std::vector<AttrRule> attributes;
+		std::string name_fmt;
+
+		size_t argidx;
+		for (argidx = 1; argidx < args.size(); argidx++) {
+			if (args[argidx] == "-setattr" && argidx+1 < args.size()) {
+				attributes.emplace_back(RTLIL::escape_id(args[++argidx]), "");
+			} else if (args[argidx] == "-formatattr" && argidx+2 < args.size()) {
+				IdString id = RTLIL::escape_id(args[++argidx]);
+				attributes.emplace_back(id, args[++argidx]);
+			} else if (args[argidx] == "-name" && argidx+1 < args.size()) {
+				name_fmt = args[++argidx];
+			} else {
+				break;
+			}
+		}
+		extra_args(args, argidx, d);
+
+		if (name_fmt.empty())
+			log_cmd_error("Argument -name required");
+
+		CellTypes ct;
+		ct.setup();
+
+		for (auto module : d->selected_modules()) {
+			for (auto cell : module->selected_cells()) {
+				std::optional<std::string> unescaped_name = format(name_fmt, cell->parameters);
+				if (!unescaped_name)
+					log_error("Formatting error when processing cell '%s' in module '%s'\n",
+							  log_id(cell), log_id(module));
+
+				IdString name = RTLIL::escape_id(unescaped_name.value());
+
+				if (d->module(name)) {
+					cell->type = name;
+					cell->parameters.clear();
+					continue;
+				}
+
+				if (!ct.cell_known(cell->type))
+					log_error("Non-internal cell type '%s' on cell '%s' in module '%s' unsupported\n",
+							  log_id(cell->type), log_id(cell), log_id(module));
+
+				Module *subm = d->addModule(name);
+				Cell *subcell = subm->addCell("$1", cell->type);
+				for (auto conn : cell->connections()) {
+					Wire *w = subm->addWire(conn.first, conn.second.size());
+					if (ct.cell_output(cell->type, w->name))
+						w->port_output = true;
+					else
+						w->port_input = true;
+					subcell->setPort(conn.first, w);
+				}
+				subcell->parameters = cell->parameters;
+				subm->fixup_ports();
+
+				for (auto rule : attributes) {
+					if (rule.value_fmt.empty()) {
+						subm->set_bool_attribute(rule.name);
+					} else {
+						std::optional<std::string> value = format(rule.value_fmt, cell->parameters);
+
+						if (!value)
+							log_error("Formatting error when processing cell '%s' in module '%s'\n",
```

Comment:
What this PR does currently works ok, but this pattern of log + log_error makes me wish for something like Result types in other languages. Ultimately what we want to do is return from `format` either a `Result(formatted_string)` or `Error(msg)` and if it's Error then return a `Error(msg with pass-level context)` from `execute`. [This](https://github.com/oktal/result) could kind of get it done, but there's no way to match on it, and `std::variant` is too annoying